File: sample-app-java/src/main/java/com/publix/grocery/pricing/WeeklyAd.java
package com.publix.grocery.pricing;

import java.math.BigDecimal;

/** A sealed hierarchy of promo types that can be attached to a SKU. */
public sealed interface WeeklyAd
    permits WeeklyAd.PercentOff, WeeklyAd.DollarOff, WeeklyAd.Bogo, WeeklyAd.BuyNForX {

    String sku();

    record PercentOff(String sku, BigDecimal percent) implements WeeklyAd {}
    record DollarOff(String sku, BigDecimal amount) implements WeeklyAd {}
    /** Buy one get one free (free item is the cheapest). */
    record Bogo(String sku) implements WeeklyAd {}
    /** Buy N units for a flat price. */
    record BuyNForX(String sku, int n, BigDecimal flatPrice) implements WeeklyAd {}
}

File: sample-app-dotnet/src/Publix.Grocery.Api/Pricing/WeeklyAd.cs
namespace Publix.Grocery.Api.Pricing;

public abstract record WeeklyAd(string Sku)
{
    public sealed record PercentOff(string Sku, decimal Percent) : WeeklyAd(Sku);
    public sealed record DollarOff(string Sku, decimal Amount) : WeeklyAd(Sku);
    /// <summary>Buy one, get one free (free item is the cheapest).</summary>
    public sealed record Bogo(string Sku) : WeeklyAd(Sku);
    /// <summary>Buy N units for a flat price.</summary>
    public sealed record BuyNForX(string Sku, int N, decimal FlatPrice) : WeeklyAd(Sku);
}
